packages/iov-core/src/jsrpc.ts

```diff
@@ -0,0 +1,242 @@
+    throw new Error("Data must be JS RPC compatible dictionary");
+  }
+
+  const id = data.id;
+  if (typeof id !== "number") {
+    return null;
+  }
+  return id;
+}
+
+export function parseJsRpcRequest(data: unknown): JsRpcRequest {
+  if (!isJsRpcCompatibleDictionary(data)) {
+    throw new Error("Data must be JS RPC compatible dictionary");
+  }
+
+  const id = parseJsRpcId(data);
+  if (id === null) {
+    throw new Error("Invalid id field");
+  }
+
+  const method = data.method;
+  if (typeof method !== "string") {
+    throw new Error("Invalid method field");
+  }
+
+  if (!isJsRpcCompatibleArray(data.params) && !isJsRpcCompatibleDictionary(data.params)) {
+    throw new Error("Invalid params field");
+  }
+
+  return {
+    id: id,
+    method: method,
+    params: data.params,
+  };
+}
+
+function parseError(error: JsRpcCompatibleDictionary): JsRpcError {
+  if (typeof error.code !== "number") {
+    throw new Error("Error property 'code' is not a number");
+  }
+
+  if (typeof error.message !== "string") {
+    throw new Error("Error property 'message' is not a string");
+  }
+
+  let maybeUndefinedData: JsRpcCompatibleValue | undefined;
+
+  if (error.data === undefined) {
+    maybeUndefinedData = undefined;
+  } else if (isJsRpcCompatibleValue(error.data)) {
+    maybeUndefinedData = error.data;
+  } else {
+    throw new Error("Error property 'data' is defined but not a JS RPC compatible value.");
+  }
+
+  return {
+    code: error.code,
+    message: error.message,
+    data: maybeUndefinedData,
+  };
+}
+
+export function parseJsRpcErrorResponse(data: unknown): JsRpcErrorResponse | undefined {
+  if (!isJsRpcCompatibleDictionary(data)) {
+    throw new Error("Data must be JS RPC compatible dictionary");
+  }
+
+  const id = parseJsRpcId(data);
+  if (id === null) {
+    throw new Error("Invalid id field");
+  }
+
+  if (typeof data.error === "undefined") {
+    return undefined;
+  }
+
+  if (!isJsRpcCompatibleDictionary(data.error)) {
+    throw new Error("Property 'error' is defined but not a JS RPC compatible dictionary");
+  }
+
+  return {
+    id: id,
+    error: parseError(data.error),
+  };
+}
+
+export function parseJsRpcResponse(data: unknown): JsRpcSuccessResponse {
+  if (!isJsRpcCompatibleDictionary(data)) {
+    throw new Error("Data must be JS RPC compatible dictionary");
+  }
+
+  const id = parseJsRpcId(data);
+  if (id === null) {
+    throw new Error("Invalid id field");
+  }
+
+  const result = data.result;
+
+  return {
+    id: id,
+    result: result,
+  };
+}
+
+export class JsRpcClient {
+  private readonly connection: SimpleMessagingConnection<JsRpcRequest, JsRpcResponse>;
+
+  constructor(connection: SimpleMessagingConnection<JsRpcRequest, JsRpcResponse>) {
+    this.connection = connection;
+  }
+
+  public async run(request: JsRpcRequest): Promise<JsRpcSuccessResponse> {
+    const filteredStream = this.connection.responseStream.filter(r => r.id === request.id);
```

Comment:
Love these two lines, to easily handle dispatching responses coming in any order. Oh, the power of streams.